From a778a273bd8ea6781914f5fbd919bcb46499c86c Mon Sep 17 00:00:00 2001
From: Jason Wang <jasowang@redhat.com>
Date: Fri, 24 Oct 2014 10:57:37 -0400
Subject: [virt] hyperv: Fix a bug in netvsc_start_xmit()

Message-id: <1414148260-34692-2-git-send-email-jasowang@redhat.com>
Patchwork-id: 98426
O-Subject: [RHEL7.1 PATCH 1/4] hyperv: Fix a bug in netvsc_start_xmit()
Bugzilla: 1156305
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Vitaly Kuznetsov <vkuznets@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

From: KY Srinivasan <kys@microsoft.com>

Bugzilla: 1156305

After the packet is successfully sent, we should not touch the skb
as it may have been freed. This patch is based on the work done by
Long Li <longli@microsoft.com>.

In this version of the patch I have fixed issues pointed out by David.
David, please queue this up for stable.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Tested-by: Long Li <longli@microsoft.com>
Tested-by: Sitsofe Wheeler <sitsofe@yahoo.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit dedb845ded56ded1c62f5398a94ffa8615d4592d)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/hyperv/netvsc_drv.c b/drivers/net/hyperv/netvsc_drv.c
index 9e67ccd..35d57f1 100644
--- a/drivers/net/hyperv/netvsc_drv.c
+++ b/drivers/net/hyperv/netvsc_drv.c
@@ -387,6 +387,7 @@ static int netvsc_start_xmit(struct sk_buff *skb, struct net_device *net)
  int  hdr_offset;
  u32 net_trans_info;
  u32 hash;
+ u32 skb_length = skb->len;
 
 
  /* We will atmost need two pages to describe the rndis
@@ -562,7 +563,7 @@ do_send:
 
 drop:
  if (ret == 0) {
-  net->stats.tx_bytes += skb->len;
+  net->stats.tx_bytes += skb_length;
   net->stats.tx_packets++;
  } else {
   kfree(packet);
-- 
1.7.1